src/module_compilation.ml

```diff
@@ -88,7 +88,12 @@ let build_cm cctx ?sandbox ?(dynlink=true) ~dep_graphs ~cm_kind (m : Module.t) =
           SC.add_rule sctx (Build.symlink ~src:in_obj_dir ~dst:in_dir))
       end;
       let opaque_arg =
-        if opaque && cm_kind = Cmi then
+        let intf_only = (* we add opaque even outside "dev" for mli only *)
+          cm_kind = Cmi
+          && not (Module.has_impl m)
+          && Ocaml_version.supports_opaque_for_mli ctx.version
```

Comment:
It feel a bit odd that `supports_opaque_for_mli` is connected to the module being `intf_only`. It would seem clearer to me to write:

```
if opaque || (intf_only && Ocaml_version.supports_opaque_for_mli ctx.version) then
```

Reply:
Yeah, I think that would be better as well.